win32/src/gballoc_ll_jemalloc.c

```diff
@@ -214,3 +218,97 @@ void gballoc_ll_print_stats(void)
     /* Codes_SRS_GBALLOC_LL_JEMALLOC_01_008: [ gballoc_ll_print_stats shall call je_malloc_stats_print and pass to it jemalloc_print_stats_callback as print callback. ]*/
     je_malloc_stats_print(jemalloc_print_stats_callback, NULL, NULL);
 }
+
+int gballoc_ll_set_decay(int64_t decay_milliseconds)
+{
+    int result = 0;
+
+    /* Codes_SRS_GBALLOC_LL_JEMALLOC_28_001: [ gballoc_ll_set_decay shall advance jemalloc's internal epoch counter. ]*/
+    if (je_mallctl("epoch", NULL, NULL, NULL, 0) != 0)
+    {
+        /* Codes_SRS_GBALLOC_LL_JEMALLOC_28_008: [ If there are any errors, gballoc_ll_set_decay shall fail and return a non-zero value. ]*/
+        LogError("failed to advance jemalloc epoch");
+        result = MU_FAILURE;
+    }
+    else
+    {
+        /* Codes_SRS_GBALLOC_LL_JEMALLOC_28_002: [ gballoc_ll_set_decay shall set the dirty decay time for new arenas to decay_milliseconds milliseconds. ]*/
+        if (je_mallctl("arenas.dirty_decay_ms", NULL, NULL, &decay_milliseconds, sizeof(decay_milliseconds)) != 0)
+        {
+            /* Codes_SRS_GBALLOC_LL_JEMALLOC_28_008: [ If there are any errors, gballoc_ll_set_decay shall fail and return a non-zero value. ]*/
+            LogError("failed to set dirty decay time");
+            result = MU_FAILURE;
+        }
+        else
+        {
+            /* Codes_SRS_GBALLOC_LL_JEMALLOC_28_003: [ gballoc_ll_set_decay shall set the muzzy decay time for new arenas to decay_milliseconds milliseconds. ]*/
+            if (je_mallctl("arenas.muzzy_decay_ms", NULL, NULL, &decay_milliseconds, sizeof(decay_milliseconds)) != 0)
+            {
+                /* Codes_SRS_GBALLOC_LL_JEMALLOC_28_008: [ If there are any errors, gballoc_ll_set_decay shall fail and return a non-zero value. ]*/
+                LogError("failed to set muzzy decay time"); 
+                result = MU_FAILURE; 
+            }
+            else
+            {
+                uint64_t narenas;
+
+                /* Codes_SRS_GBALLOC_LL_JEMALLOC_28_004: [ gballoc_ll_set_decay shall fetch the number of existing jemalloc arenas. ]*/
+                if (je_mallctl("arenas.narenas", &narenas, &(size_t){sizeof(narenas)}, NULL, 0) != 0)
+                {
+                    /* Codes_SRS_GBALLOC_LL_JEMALLOC_28_008: [ If there are any errors, gballoc_ll_set_decay shall fail and return a non-zero value. ]*/
+                    LogError("failed to get number of arenas");
+                    result = MU_FAILURE;
+                }
+                else
+                {
+                    char command[64];
+
+                    /* Codes_SRS_GBALLOC_LL_JEMALLOC_28_005: [ For each existing arena: ]*/
+                    for (uint64_t i = 0; i < narenas; i++) {
+                        int snprintf_result = snprintf(command, sizeof(command), "arena.%" PRIu64 ".dirty_decay_ms", i);
+
+                        if (snprintf_result < 0)
```

Comment:
>snprintf_result < 0 [](http://example.com/codeflow?start=28&length=19)

What happens is line 264 is not so generous with the number of bytes and uses... hmmm.. 10 instead of 64?

C17 standard says `The snprintf function returns the number of characters that would have been written had n been
 sufficiently large, not counting the terminating null character, or a negative value if an encoding
 error occurred. Thus, the null-terminated output has been completely written if and only if the
 returned value is nonnegative and less than n.`

Reply:
I have added the extra check for snprintf. Do you want me to reduce the command array size to something lesser than 64? Since this won't be a frequent operation, I kept it a bit higher.